processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -1007,18 +862,41 @@ public Metadata apply(IndexableAdapter input)
     );
 
     indexIO.getDefaultIndexIOHandler().convertV8toV9(v8OutDir, outDir, indexSpec);
+    ioPeon.cleanup();
     FileUtils.deleteDirectory(v8OutDir);
 
     return outDir;
   }
 
+  protected DimensionHandler[] makeDimensionHandlers(final List<String> mergedDimensions, final List<ColumnCapabilitiesImpl> dimCapabilities)
+  {
+    final DimensionHandler[] handlers = new DimensionHandler[mergedDimensions.size()];
+    for (int i = 0; i < mergedDimensions.size(); i++) {
+      ColumnCapabilities capabilities = dimCapabilities.get(i);
+      String dimName = mergedDimensions.get(i);
+      if (capabilities.getType() == ValueType.STRING && capabilities.isDictionaryEncoded()) {
+        handlers[i] = new StringDimensionHandler(dimName);
+      }
+
+      if (capabilities.getType() == ValueType.LONG && !capabilities.isDictionaryEncoded()) {
+        handlers[i] = new LongDimensionHandler(dimName);
+      }
+
+      if (capabilities.getType() == ValueType.FLOAT && !capabilities.isDictionaryEncoded()) {
+        handlers[i] = new FloatDimensionHandler(dimName);
+      }
+    }
```

Comment:
@fjy I've moved this capabilities -> DimensionHandler code to a static method on a util class, DimensionHandlerUtil.

It will throw an IAE if capabilities type is not String/Long/Float or if the dictionary encoding/bitmap indexing settings are incorrect for a given type